vcr/issuer/statuslist2021_store.go

```diff
@@ -0,0 +1,221 @@
+	"github.com/nuts-foundation/nuts-node/vcr/types"
+	"strconv"
+	"sync"
+
+	ssi "github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/vcr/credential"
+	"github.com/nuts-foundation/nuts-node/vcr/credential/statuslist2021"
+	"github.com/nuts-foundation/nuts-node/vdr/didweb"
+)
+
+// errNotFound wraps types.ErrNotFound to clarify which credential is not found
+var errNotFound = fmt.Errorf("status list: %w", types.ErrNotFound)
+
+// errUnsupportedPurpose limits current usage to 'revocation'
+var errUnsupportedPurpose = errors.New("status list: purpose not supported")
+
+// errNotFound wraps types.ErrRevoked to clarify the source of the error
+var errRevoked = fmt.Errorf("status list: %w", types.ErrRevoked)
+
+type StatusPurpose string
+
+const (
+	StatusPurposeRevocation = "revocation"
+	statusPurposeSuspension = "suspension" // currently not supported
+)
+
+// StatusList2021Store keeps track of the number of credentials with a credential status per issuer, and allows revoking
+// them by setting the relevant bit on the StatusList.
+type StatusList2021Store interface {
+	// CredentialSubject creates and signs a StatusList2021Credential that is hosted on issuerURL/status/page.
+	CredentialSubject(ctx context.Context, issuer did.DID, page string) (*credential.StatusList2021CredentialSubject, error)
+	// Create a credential.StatusList2021Entry that can be added to the credentialStatus of a VC.
+	// The corresponding credential will have a gap in the bitstring if the returned entry does not make it into a credential.
+	Create(ctx context.Context, issuer did.DID, purpose StatusPurpose) (*credential.StatusList2021Entry, error)
+	// Revoke by adding StatusList2021Entry to the list of revocations.
+	// The credentialID is only used to allow reverse search of revocations, its issuer is NOT compared to the entry issuer.
+	// Returns types.ErrRevoked if already revoked, or types.ErrNotFound when the entry.StatusListCredential is unknown.
+	Revoke(ctx context.Context, credentialID ssi.URI, entry credential.StatusList2021Entry) error
+
+	// individual statuses should be derived from the StatusListCredential, not inspected here.
+}
+
+// lastIndex tracks the last Page + StatusListIndex combo issued for an Issuer
+type lastIndex struct {
+	Issuer          string // did:web:example.com:iam:id
+	Page            int    // >= 1
+	StatusListIndex int    // 0 <= statusListIndex < statuslist2021.MaxBitstringIndex
+}
+
+// revocation exists if the index on the statusListCredential is set
+type revocation struct {
+	StatusListCredential string // https://example.com/iam/id/status/page
+	CredentialID         string // did:web:example.com:iam:id#unique-identifier
+	Page                 int    // >= 1
+	StatusListIndex      int    // 0 <= statusListIndex < statuslist2021.MaxBitstringIndex
+}
+
+var _ StatusList2021Store = (*statusListMemoryStore)(nil)
+
+func newStatusListMemoryStore() *statusListMemoryStore {
```

Comment:
I'd like to see this in-memory store removed as soon as we have a SQL implementation. In test, SQLite is used which is almost as easy as this in-memory store.

Reply:
That is the plan. And that is when I will add tests for the store.

Reply:
#2784